File: chrome/browser/android/password_authentication_manager.cc
// Copyright 2014 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include "chrome/browser/android/password_authentication_manager.h"

#include "chrome/browser/android/tab_android.h"
#include "jni/PasswordAuthenticationManager_jni.h"

namespace {

class PasswordAuthenticationCallback {
 public:
  explicit PasswordAuthenticationCallback(
      const base::Closure& success_callback) {
    success_callback_ = success_callback;
  }

  void OnResult(bool result) {
    if (result)
      success_callback_.Run();
    delete this;
  }

 private:
  virtual ~PasswordAuthenticationCallback() {}

  base::Closure success_callback_;
};

}  // namespace

PasswordAuthenticationManager::PasswordAuthenticationManager() {
}

PasswordAuthenticationManager::~PasswordAuthenticationManager() {
}

bool PasswordAuthenticationManager::RegisterPasswordAuthenticationManager(
    JNIEnv* env) {
  return RegisterNativesImpl(env);
}

void PasswordAuthenticationManager::AuthenticatePasswordAutofill(
    content::WebContents* web_contents,
    const base::Closure& success_callback) {
  TabAndroid* tab = TabAndroid::FromWebContents(web_contents);
  if (!tab)
    return;

  JNIEnv* env = base::android::AttachCurrentThread();
  PasswordAuthenticationCallback* auth_callback =
      new PasswordAuthenticationCallback(success_callback);
  Java_PasswordAuthenticationManager_requestAuthentication(
      env,
      tab->GetJavaObject().obj(),
      Java_PasswordAuthenticationCallback_create(
          env,
          reinterpret_cast<intptr_t>(auth_callback)).obj());
}

// static
void OnResult(JNIEnv* env,
              jclass jcaller,
              jlong callback_ptr,
              jboolean authenticated) {
  PasswordAuthenticationCallback* callback =
      reinterpret_cast<PasswordAuthenticationCallback*>(callback_ptr);
  callback->OnResult(authenticated);
}

File: chrome/android/java/src/org/chromium/chrome/browser/password_manager/PasswordAuthenticationManager.java
// Copyright 2014 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

package org.chromium.chrome.browser.password_manager;

import org.chromium.base.CalledByNative;
import org.chromium.chrome.browser.Tab;

/**
 * Allows embedders to authenticate the usage of passwords.
 */
public class PasswordAuthenticationManager {

    /**
     * The delegate that allows embedders to control the authentication of passwords.
     */
    public interface PasswordAuthenticationDelegate {
        /**
         * @return Whether password authentication is enabled.
         */
        boolean isPasswordAuthenticationEnabled();

        /**
         * Requests password authentication be presented for the given tab.
         * @param tab The tab containing the protected password.
         * @param callback The callback to be triggered on authentication result.
         */
        void requestAuthentication(Tab tab, PasswordAuthenticationCallback callback);

        /**
         * @return The message to be displayed in the save password infobar that will allow
         *         the user to opt-in to additional password authentication.
         */
        String getPasswordProtectionString();
    }

    /**
     * The callback to be triggered on success or failure of the password authentication.
     */
    public static class PasswordAuthenticationCallback {
        private long mNativePtr;

        @CalledByNative("PasswordAuthenticationCallback")
        private static PasswordAuthenticationCallback create(long nativePtr) {
            return new PasswordAuthenticationCallback(nativePtr);
        }

        private PasswordAuthenticationCallback(long nativePtr) {
            mNativePtr = nativePtr;
        }

        /**
         * Called upon authentication results to allow usage of the password or not.
         * @param authenticated Whether the authentication was successful.
         */
        public final void onResult(boolean authenticated) {
            if (mNativePtr == 0) {
                assert false : "Can not call onResult more than once per callback.";
                return;
            }
            nativeOnResult(mNativePtr, authenticated);
            mNativePtr = 0;
        }
    }

    private static class DefaultPasswordAuthenticationDelegate
            implements PasswordAuthenticationDelegate {
        @Override
        public boolean isPasswordAuthenticationEnabled() {
            return false;
        }

        @Override
        public void requestAuthentication(Tab tab, PasswordAuthenticationCallback callback) {
            callback.onResult(true);
        }

        @Override
        public String getPasswordProtectionString() {
            return "";
        }
    }

    private static PasswordAuthenticationDelegate sDelegate;

    private PasswordAuthenticationManager() {}

    private static PasswordAuthenticationDelegate getDelegate() {
        if (sDelegate == null) {
            sDelegate = new DefaultPasswordAuthenticationDelegate();
        }
        return sDelegate;
    }

    /**
     * Sets the password authentication delegate to be used.
     */
    public static void setDelegate(PasswordAuthenticationDelegate delegate) {
        sDelegate = delegate;
    }

    /**
     * @return Whether password authentication is enabled.
     */
    public static boolean isPasswordAuthenticationEnabled() {
        return getDelegate().isPasswordAuthenticationEnabled();
    }

    /**
     * Requests password authentication be presented for the given tab.
     * @param tab The tab containing the protected password.
     * @param callback The callback to be triggered on authentication result.
     */
    @CalledByNative
    public static void requestAuthentication(
            Tab tab, PasswordAuthenticationCallback callback) {
        getDelegate().requestAuthentication(tab, callback);
    }

    /**
     * @return The message to be displayed in the save password infobar that will allow the user
     *         to opt-in to additional password authentication.
     */
    public static String getPasswordProtectionString() {
        return getDelegate().getPasswordProtectionString();
    }

    private static native void nativeOnResult(long callbackPtr, boolean authenticated);
}
